File: src/transpiler.ts
import { exec } from "node:child_process";
import fs from "node:fs/promises";
import path from "node:path";
import { promisify } from "node:util";
import {
  getAllKeywords,
  getFunctionKeywords,
  getMultiWordKeywords,
} from "./keywords.js";

export class NullScriptTranspileError extends Error {
  constructor(
    message: string,
    public filePath?: string,
    public line?: number,
    public column?: number,
  ) {
    super(message);
    this.name = "NullScriptTranspileError";
  }
}

export class NullScriptSyntaxError extends NullScriptTranspileError {
  constructor(
    message: string,
    filePath?: string,
    line?: number,
    column?: number,
  ) {
    super(message, filePath, line, column);
    this.name = "NullScriptSyntaxError";
  }
}

export class NullScriptTypeError extends NullScriptTranspileError {
  constructor(
    message: string,
    filePath?: string,
    line?: number,
    column?: number,
  ) {
    super(message, filePath, line, column);
    this.name = "NullScriptTypeError";
  }
}

const execAsync = promisify(exec);

// Validation functions
function validateNullScriptSyntax(source: string, filePath?: string): void {
  const lines = source.split("\n");

  // Check for common NullScript syntax issues
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    const lineNumber = i + 1;

    // Skip empty lines and comments
    if (!line || line.startsWith("//") || line.startsWith("/*")) continue;

    // Check for invalid keywords (not in our keyword list but look like they should be)
    const invalidKeywordPatterns = [
      /\b(function\s+\w+\s*\()/, // using 'function' instead of 'feels'
      /\b(const\s+\w+)/, // using 'const' instead of 'definitely'
      /\b(let\s+\w+)/, // using 'let' instead of 'maybe'
      /\b(var\s+\w+)/, // using 'var' instead of 'mayhap'
      /\b(if\s*\()/, // using 'if' instead of 'checkthis'
      /\b(else\s+)/, // using 'else' instead of 'orelse'
      /\b(return\s+)/, // using 'return' instead of 'pls'
      /\b(true)\b/, // using 'true' instead of 'fr'
      /\b(false)\b/, // using 'false' instead of 'cap'
      /\b(null)\b/, // using 'null' instead of 'nocap'
      /\b(undefined)\b/, // using 'undefined' instead of 'ghost'
      /\b(interface\s+\w+)/, // using 'interface' instead of 'vibes'
      /\b(type\s+\w+)/, // using 'type' instead of 'vibe'
      /\b(class\s+\w+)/, // using 'class' instead of 'bigbrain'
      /\b(try\s*\{)/, // using 'try' instead of 'oops' or 'oop'
      /\b(catch\s*\()/, // using 'catch' instead of 'mybad'
      /\b(finally\s*\{)/, // using 'finally' instead of 'anyway'
    ];

    for (const pattern of invalidKeywordPatterns) {
      if (pattern.test(line)) {
        throw new NullScriptSyntaxError(
          `Invalid syntax on line ${lineNumber}: You're using standard TypeScript/JavaScript syntax instead of NullScript keywords.\n💡 Run 'nullc keywords' to see the correct NullScript syntax.`,
          filePath,
          lineNumber,
        );
      }
    }

    // Check for completely invalid keywords
    const unknownKeywordPattern = /^(\w+)\s+\w+\s*=/;
    const match = line.match(unknownKeywordPattern);
    if (match) {
      const keyword = match[1];
      // Check if this keyword is not in our valid NullScript keywords
      const allKeywords = Object.values(getAllKeywords());
      const allKeys = Object.keys(getAllKeywords());

      if (
        !allKeys.includes(keyword) &&
        !["export", "import", "from", "as"].includes(keyword)
      ) {
        throw new NullScriptSyntaxError(
          `Unknown keyword '${keyword}' on line ${lineNumber}.\n💡 Use valid NullScript keywords. Run 'nullc keywords' to see all available options.`,
          filePath,
          lineNumber,
        );
      }
    }
  }
}

// TypeScript error patterns and their NullScript equivalents
const ERROR_MAPPINGS = {
  // Common syntax errors
  "Cannot find name 'feels'": {
    type: "syntax",
    message:
      "Invalid function declaration. Use 'feels' followed by a function name.",
    suggestion: "Example: feels myFunction() { ... }",
  },
  "Cannot find name 'definitely'": {
    type: "syntax",
    message: "Invalid variable declaration. Use 'definitely' for constants.",
    suggestion: "Example: definitely myVar = 'value'",
  },
  "Cannot find name 'maybe'": {
    type: "syntax",
    message:
      "Invalid variable declaration. Use 'maybe' for variables that can change.",
    suggestion: "Example: maybe myVar = 'value'",
  },
  "Cannot find name 'checkthis'": {
    type: "syntax",
    message:
      "Invalid conditional statement. Use 'checkthis' for if statements.",
    suggestion: "Example: checkthis (condition) { ... }",
  },
  "Cannot find name 'orelse'": {
    type: "syntax",
    message: "Invalid else statement. Use 'orelse' for else clauses.",
    suggestion: "Example: checkthis (condition) { ... } orelse { ... }",
  },
  "Cannot find name 'pls'": {
    type: "syntax",
    message: "Invalid return statement. Use 'pls' to return values.",
    suggestion: "Example: pls myValue",
  },
  "Cannot find name 'fr'": {
    type: "syntax",
    message: "Invalid boolean value. Use 'fr' for true.",
    suggestion: "Example: definitely isValid = fr",
  },
  "Cannot find name 'cap'": {
    type: "syntax",
    message: "Invalid boolean value. Use 'cap' for false.",
    suggestion: "Example: definitely isValid = cap",
  },
  "Cannot find name 'nocap'": {
    type: "syntax",
    message: "Invalid null value. Use 'nocap' for null.",
    suggestion: "Example: definitely value = nocap",
  },
  "Cannot find name 'ghost'": {
    type: "syntax",
    message: "Invalid undefined value. Use 'ghost' for undefined.",
    suggestion: "Example: definitely value = ghost",
  },
  "Cannot find name 'vibes'": {
    type: "syntax",
    message: "Invalid interface declaration. Use 'vibes' to define interfaces.",
    suggestion: "Example: vibes MyInterface { ... }",
  },
  "Cannot find name 'vibe'": {
    type: "syntax",
    message: "Invalid type alias. Use 'vibe' to define type aliases.",
    suggestion: "Example: vibe MyType = string | number",
  },
  "Cannot find name 'bigbrain'": {
    type: "syntax",
    message: "Invalid class declaration. Use 'bigbrain' to define classes.",
    suggestion: "Example: bigbrain MyClass { ... }",
  },
  "Unexpected token": {
    type: "syntax",
    message:
      "Syntax error in NullScript code. Check for missing keywords or incorrect syntax.",
    suggestion:
      "Make sure you're using NullScript keywords correctly. Run 'nullc keywords' to see all available keywords.",
  },
  "Declaration or statement expected": {
    type: "syntax",
    message: "Invalid statement. Check your NullScript syntax.",
    suggestion: "Make sure you're using proper NullScript keywords and syntax.",
  },
  "Function implementation is missing": {
    type: "syntax",
    message:
      "Function body is missing. Add implementation after your function declaration.",
    suggestion: "Example: feels myFunction() { /* your code here */ }",
  },
  "Unexpected keyword or identifier": {
    type: "syntax",
    message:
      "Invalid NullScript syntax. You're using an undefined keyword or incorrect syntax.",
    suggestion:
      "Check that you're using valid NullScript keywords. Run 'nullc keywords' to see all available options.",
  },
};

export function parseTypeScriptError(
  error: string,
  filePath?: string,
): NullScriptTranspileError {
  const lines = error.split("\n");
  let line: number | undefined;
  let column: number | undefined;
  let errorMessage = error;

  // Extract line and column information from TypeScript error
  // Handle format like "file.ts:12:1 - error TS1434:"
  const locationMatch = error.match(
    /(\w+\.ts):(\d+):(\d+)\s*-\s*error|:(\d+):(\d+)/,
  );
  if (locationMatch) {
    line = parseInt(locationMatch[2] || locationMatch[4]);
    column = parseInt(locationMatch[3] || locationMatch[5]);
  }

  // Find the core error message(s) - handle multiple TypeScript errors
  const errorLines = lines.filter((line) => line.includes("error TS"));
  if (errorLines.length > 0) {
    // Take the first meaningful error
    const firstError = errorLines[0];
    const errorMatch = firstError.match(/error TS\d+: (.+)/);
    if (errorMatch) {
      errorMessage = errorMatch[1];
    }
  }

  // If no TypeScript errors found, look for other common patterns
  if (!errorLines.length) {
    // Look for compilation failure patterns
    const compilationError = lines.find(
      (line) =>
        line.includes("Cannot find name") ||
        line.includes("Unexpected token") ||
        line.includes("Declaration or statement expected"),
    );
    if (compilationError) {
      errorMessage = compilationError.trim();
    }
  }

  // Check for known error patterns
  for (const [pattern, mapping] of Object.entries(ERROR_MAPPINGS)) {
    if (errorMessage.includes(pattern)) {
      const customMessage = `${mapping.message}\n💡 ${mapping.suggestion}`;

      if (mapping.type === "syntax") {
        return new NullScriptSyntaxError(customMessage, filePath, line, column);
      } else {
        return new NullScriptTypeError(customMessage, filePath, line, column);
      }
    }
  }

  // Fallback: clean up the TypeScript error message
  const cleanMessage = errorMessage
    .replace(/error TS\d+:\s*/, "")
    .replace(/\s+/g, " ")
    .replace(/^\s*at\s+.*$/gm, "") // Remove stack trace lines
    .replace(/Command failed:.*$/m, "") // Remove command failure messages
    .replace(/\(node:\d+\).*$/gm, "") // Remove node warnings
    .split("\n")
    .filter((line) => line.trim().length > 0)
    .slice(0, 3) // Take first 3 meaningful lines
    .join("\n")
    .trim();

  return new NullScriptTranspileError(
    `Transpilation error: ${cleanMessage}\n💡 This might be due to incorrect NullScript syntax. Run 'nullc keywords' to see available keywords.`,
    filePath,
    line,
    column,
  );
}

export function formatNullScriptError(error: NullScriptTranspileError): string {
  let output = `❌ ${error.name}`;

  if (error.filePath) {
    output += ` in ${path.basename(error.filePath)}`;
  }

  if (error.line !== undefined) {
    output += `:${error.line}`;
    if (error.column !== undefined) {
      output += `:${error.column}`;
    }
  }

  output += `\n\n${error.message}`;

  return output;
}

export type TranspileOptions = {
  outputFormat?: "ts" | "js";
  outDir?: string;
  preserveComments?: boolean;
  skipTypeCheck?: boolean;
};

export function transpileNullScript(
  source: string,
  options: TranspileOptions = {},
): string {
  let output = source;

  for (const [alias, tsKeyword] of Object.entries(getFunctionKeywords())) {
    if (alias.includes("async")) {
      output = output.replace(
        new RegExp(`\\b${alias}\\s+([a-zA-Z_$][\\w$]*)`, "g"),
        `${tsKeyword} $1`,
      );
    } else {
      output = output.replace(
        new RegExp(
          `\\b${alias}\\s+([a-zA-Z_$][\\w$]*)\\s*(?:<[^>]*>)?\\s*\\(`,
          "g",
        ),
        (match, functionName) => {
          const lines = output.split("\n");
          const currentLineIndex =
            output.substring(0, output.indexOf(match)).split("\n").length - 1;
          const currentLine = lines[currentLineIndex];
          const indent = currentLine.match(/^(\s*)/)?.[1] || "";

          const restOfMatch = match.substring(
            alias.length + 1 + functionName.length,
          );

          if (indent.length > 0) {
            return `${functionName}${restOfMatch}`;
          } else {
            return `${tsKeyword} ${functionName}${restOfMatch}`;
          }
        },
      );

      output = output.replace(
        new RegExp(`\\b${alias}\\s*\\(`, "g"),
        `${tsKeyword}(`,
      );
    }
  }

  for (const [alias, tsKeyword] of Object.entries(getAllKeywords())) {
    if (alias === "feels" || alias === "feels async") continue;

    if (alias === "remove") {
      output = output.replace(
        new RegExp(
          `\\bremove\\s+([a-zA-Z_$][\\w$]*(?:\\.[a-zA-Z_$][\\w$]*)*(?:\\[[^\\]]+\\])?)\\b`,
          "g",
        ),
        `delete $1`,
      );
    } else {
      const regex = new RegExp(`\\b${alias}\\b`, "g");
      output = output.replace(regex, tsKeyword);
    }
  }

  for (const [alias, tsKeyword] of Object.entries(getMultiWordKeywords())) {
    output = output.replace(
      new RegExp(`\\b${alias}\\s+`, "g"),
      `${tsKeyword} `,
    );
  }

  return output;
}

export async function transpileFile(
  inputPath: string,
  outputPath: string,
  options: TranspileOptions = {},
): Promise<string> {
  const source = await fs.readFile(inputPath, "utf8");

  // Validate NullScript syntax before transpiling
  validateNullScriptSyntax(source, inputPath);

  const transpiled = transpileNullScript(source, options);

  await fs.mkdir(path.dirname(outputPath), { recursive: true });
  await fs.writeFile(outputPath, transpiled, "utf8");

  return transpiled;
}

export async function transpileToJs(
  nsPath: string,
  jsPath: string,
  options: TranspileOptions = {},
): Promise<void> {
  const tsPath = nsPath.replace(/\.ns$/, ".ts");

  try {
    await transpileFile(nsPath, tsPath, { ...options, outputFormat: "ts" });
  } catch (error) {
    // If transpilation to .ts fails, make sure no .ts file is left behind
    await fs.unlink(tsPath).catch(() => {}); // ignore errors if file doesn't exist
    throw error;
  }

  const tempDir = path.dirname(tsPath);
  const tsConfigPath = path.join(tempDir, "tsconfig.json");
  const tsConfig = {
    compilerOptions: {
      target: "ES2022",
      module: "ES2022",
      moduleResolution: "node",
      outDir: path.dirname(jsPath),
      rootDir: tempDir,
      esModuleInterop: true,
      allowSyntheticDefaultImports: true,
      skipLibCheck: true,
      noEmit: false,
    },
    include: [path.basename(tsPath)],
  };

  let configCreated = false;

  try {
    await fs.writeFile(tsConfigPath, JSON.stringify(tsConfig, null, 2));
    configCreated = true;

    const tscCommand = options.skipTypeCheck
      ? `npx tsc --noCheck --project "${tsConfigPath}"`
      : `npx tsc --project "${tsConfigPath}"`;

    await execAsync(tscCommand, { encoding: "utf8" });

    if (options.outputFormat === "js") {
      await fs.unlink(tsPath).catch(() => {});
    }
  } catch (error) {
    // Extract the actual error output from the exec error
    let errorOutput = "";

    if (error && typeof error === "object") {
      // TypeScript errors are typically in stdout, not stderr
      // Check stdout first for TypeScript error patterns, then stderr
      const stdout = String((error as any).stdout || "");
      const stderr = String((error as any).stderr || "");

      if (
        stdout &&
        (stdout.includes("error TS") ||
          (stdout.includes("Found") && stdout.includes("error")))
      ) {
        errorOutput = stdout;
      } else if (stderr && stderr.includes("error")) {
        errorOutput = stderr;
      } else if ("message" in error && error.message) {
        errorOutput = String(error.message);
      } else {
        errorOutput = stdout || stderr || String(error);
      }
    } else {
      errorOutput = String(error);
    }

    const customError = parseTypeScriptError(errorOutput, nsPath);
    throw customError;
  } finally {
    if (configCreated) {
      await fs.unlink(tsConfigPath).catch(() => {});
    }
  }
}

export async function buildDirectory(
  inputDir: string,
  outputDir: string,
  options: TranspileOptions = {},
): Promise<string[]> {
  const { glob } = await import("glob");
  const nsFiles = await glob("**/*.ns", { cwd: inputDir, absolute: true });

  const outputs: string[] = [];

  for (const nsFile of nsFiles) {
    const relativePath = path.relative(inputDir, nsFile);
    const outputExt = options.outputFormat === "js" ? ".js" : ".ts";
    const outputPath = path.join(
      outputDir,
      relativePath.replace(/\.ns$/, outputExt),
    );

    if (options.outputFormat === "js") {
      await transpileToJs(nsFile, outputPath, options);
    } else {
      await transpileFile(nsFile, outputPath, options);
    }

    outputs.push(outputPath);
  }

  return outputs;
}

File: src/transpiler.rs
use crate::errors::*;
use crate::keywords::NullScriptKeywords;
use regex::Regex;
use std::path::{Path, PathBuf};
use std::process::Command;
use tokio::fs;
use walkdir::WalkDir;

#[derive(Debug, Clone)]
pub struct TranspileOptions {
    pub output_format: OutputFormat,
    pub skip_type_check: bool,
}

#[derive(Debug, Clone, PartialEq)]
pub enum OutputFormat {
    TypeScript,
    JavaScript,
}

impl Default for TranspileOptions {
    fn default() -> Self {
        Self {
            output_format: OutputFormat::TypeScript,
            skip_type_check: false,
        }
    }
}

pub struct NullScriptTranspiler {
    keywords: NullScriptKeywords,
}

impl Default for NullScriptTranspiler {
    fn default() -> Self {
        Self::new()
    }
}

impl NullScriptTranspiler {
    pub fn new() -> Self {
        Self {
            keywords: NullScriptKeywords::new(),
        }
    }

    /// Validate NullScript syntax before transpiling
    pub fn validate_syntax(&self, source: &str, file_path: Option<&Path>) -> Result<(), NullScriptError> {
        let lines: Vec<&str> = source.split('\n').collect();

        for (i, line) in lines.iter().enumerate() {
            let line = line.trim();
            let line_number = i as u32 + 1;

            // Skip empty lines and comments
            if line.is_empty() || line.starts_with("//") || line.starts_with("/*") {
                continue;
            }

            // Check for invalid keywords (using standard TS/JS instead of NullScript)
            let invalid_patterns = vec![
                (r"\b(function\s+\w+\s*\()", "using 'function' instead of 'feels'"),
                (r"\b(const\s+\w+)", "using 'const' instead of 'definitely'"),
                (r"\b(let\s+\w+)", "using 'let' instead of 'maybe'"),
                (r"\b(var\s+\w+)", "using 'var' instead of 'mayhap'"),
                (r"\b(if\s*\()", "using 'if' instead of 'checkthis'"),
                (r"\b(else\s+)", "using 'else' instead of 'orelse'"),
                (r"\b(return\s+)", "using 'return' instead of 'pls'"),
                (r"\b(true)\b", "using 'true' instead of 'fr'"),
                (r"\b(false)\b", "using 'false' instead of 'cap'"),
                (r"\b(null)\b", "using 'null' instead of 'nocap'"),
                (r"\b(undefined)\b", "using 'undefined' instead of 'ghost'"),
                (r"\b(interface\s+\w+)", "using 'interface' instead of 'vibes'"),
                (r"\b(type\s+\w+)", "using 'type' instead of 'vibe'"),
                (r"\b(class\s+\w+)", "using 'class' instead of 'bigbrain'"),
                (r"\b(try\s*\{)", "using 'try' instead of 'oops' or 'oop'"),
                (r"\b(catch\s*\()", "using 'catch' instead of 'mybad'"),
                (r"\b(finally\s*\{)", "using 'finally' instead of 'anyway'"),
            ];

            for (pattern, _description) in invalid_patterns {
                let regex = Regex::new(pattern)?;
                if regex.is_match(line) {
                    let message = format!(
                        "Invalid syntax on line {}: You're using standard TypeScript/JavaScript syntax instead of NullScript keywords.\n💡 Run 'nullc keywords' to see the correct NullScript syntax.",
                        line_number
                    );
                    return Err(NullScriptError::Syntax(
                        NullScriptSyntaxError::with_location(
                            message,
                            file_path.map(|p| p.to_path_buf()),
                            Some(line_number),
                            None,
                        )
                    ));
                }
            }

            // Check for unknown keywords
            let unknown_keyword_regex = Regex::new(r"^(\w+)\s+\w+\s*=")?;
            if let Some(captures) = unknown_keyword_regex.captures(line) {
                if let Some(keyword_match) = captures.get(1) {
                    let keyword = keyword_match.as_str();
                    let all_keywords = self.keywords.get_all_keywords();
                    let valid_keywords = ["export", "import", "from", "as"];

                    if !all_keywords.contains_key(keyword) && !valid_keywords.contains(&keyword) {
                        let message = format!(
                            "Unknown keyword '{}' on line {}.\n💡 Use valid NullScript keywords. Run 'nullc keywords' to see all available options.",
                            keyword, line_number
                        );
                        return Err(NullScriptError::Syntax(
                            NullScriptSyntaxError::with_location(
                                message,
                                file_path.map(|p| p.to_path_buf()),
                                Some(line_number),
                                None,
                            )
                        ));
                    }
                }
            }
        }

        Ok(())
    }

    /// Transpile NullScript source code to TypeScript/JavaScript
    pub fn transpile(&self, source: &str, _options: &TranspileOptions) -> Result<String, NullScriptError> {
        let mut output = source.to_string();

        // Handle function declarations first
        for (alias, ts_keyword) in self.keywords.get_function_keywords() {
            if alias.contains("async") {
                let regex = Regex::new(&format!(r"\b{}\s+([a-zA-Z_$][\w$]*)", regex::escape(alias)))?;
                output = regex.replace_all(&output, format!("{} $1", ts_keyword)).to_string();
            } else {
                // Handle function declarations more carefully
                let regex = Regex::new(&format!(
                    r"\b{}\s+([a-zA-Z_$][\w$]*)\s*(?:<[^>]*>)?\s*\(",
                    regex::escape(alias)
                ))?;

                                output = regex.replace_all(&output, |caps: &regex::Captures| {
                    let _function_name = &caps[1];
                    let lines: Vec<&str> = output.split('\n').collect();
                    let match_pos = caps.get(0).unwrap().start();
                    let current_line_index = output[..match_pos].matches('\n').count();

                    if current_line_index < lines.len() {
                        let current_line = lines[current_line_index];
                        let indent = current_line.chars()
                            .take_while(|&c| c.is_whitespace())
                            .collect::<String>();

                        if !indent.is_empty() {
                            // Method declaration (indented)
                            caps[0].replace(alias, "").trim_start().to_string()
                        } else {
                            // Top-level function declaration
                            format!("{} {}", ts_keyword, caps[0].replace(alias, "").trim_start())
                        }
                    } else {
                        format!("{} {}", ts_keyword, caps[0].replace(alias, "").trim_start())
                    }
                }).to_string();

                // Handle anonymous functions
                let anon_regex = Regex::new(&format!(r"\b{}\s*\(", regex::escape(alias)))?;
                output = anon_regex.replace_all(&output, format!("{}(", ts_keyword)).to_string();
            }
        }

        // Handle regular keywords (excluding function keywords)
        let all_keywords = self.keywords.get_all_keywords();
        for (alias, ts_keyword) in all_keywords {
            if alias == "feels" || alias == "feels async" {
                continue; // Already handled above
            }

            if alias == "remove" {
                // Special handling for delete operator
                let regex = Regex::new(r"\bremove\s+([a-zA-Z_$][\w$]*(?:\.[a-zA-Z_$][\w$]*)*(?:\[[^\]]+\])?)\b")?;
                output = regex.replace_all(&output, "delete $1").to_string();
            } else {
                let regex = Regex::new(&format!(r"\b{}\b", regex::escape(&alias)))?;
                output = regex.replace_all(&output, ts_keyword.as_str()).to_string();
            }
        }

        // Handle multi-word keywords
        for (alias, ts_keyword) in self.keywords.get_multi_word_keywords() {
            let regex = Regex::new(&format!(r"\b{}\s+", regex::escape(alias)))?;
            output = regex.replace_all(&output, format!("{} ", ts_keyword)).to_string();
        }

        Ok(output)
    }

    /// Transpile a single file
    pub async fn transpile_file(
        &self,
        input_path: &Path,
        output_path: &Path,
        options: &TranspileOptions,
    ) -> Result<String, NullScriptError> {
        let source = fs::read_to_string(input_path).await?;

        // Validate syntax before transpiling
        self.validate_syntax(&source, Some(input_path))?;

        let transpiled = self.transpile(&source, options)?;

        // Create output directory if it doesn't exist
        if let Some(parent) = output_path.parent() {
            fs::create_dir_all(parent).await?;
        }

        fs::write(output_path, &transpiled).await?;

        Ok(transpiled)
    }

        /// Transpile to JavaScript via TypeScript
    pub async fn transpile_to_js(
        &self,
        ns_path: &Path,
        js_path: &Path,
        options: &TranspileOptions,
    ) -> Result<(), NullScriptError> {
        // Create a temporary directory for compilation
        let temp_dir = std::env::temp_dir().join("nullscript-temp");
        fs::create_dir_all(&temp_dir).await?;

        let ts_filename = ns_path.file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("temp")
            .to_string() + ".ts";

        let temp_ts_path = temp_dir.join(&ts_filename);

        // First transpile to TypeScript in temp directory
        let ts_options = TranspileOptions {
            output_format: OutputFormat::TypeScript,
            ..options.clone()
        };

        match self.transpile_file(ns_path, &temp_ts_path, &ts_options).await {
            Ok(_) => {},
            Err(e) => {
                // Clean up temp directory
                let _ = fs::remove_dir_all(&temp_dir).await;
                return Err(e);
            }
        }

        let tsconfig_path = temp_dir.join("tsconfig.json");
        let js_filename = ns_path.file_stem()
            .and_then(|s| s.to_str())
            .unwrap_or("temp")
            .to_string() + ".js";
        let temp_js_path = temp_dir.join(&js_filename);

        let tsconfig = serde_json::json!({
            "compilerOptions": {
                "target": "ES2022",
                "module": "ES2022",
                "moduleResolution": "node",
                "outDir": ".",
                "esModuleInterop": true,
                "allowSyntheticDefaultImports": true,
                "skipLibCheck": true,
                "noEmit": false,
            },
            "include": [ts_filename]
        });

        // Create temporary tsconfig.json
        fs::write(&tsconfig_path, serde_json::to_string_pretty(&tsconfig)?).await?;

        // Run TypeScript compiler in the temp directory
        let tsc_args = if options.skip_type_check {
            vec!["--noCheck", "--project", "tsconfig.json"]
        } else {
            vec!["--project", "tsconfig.json"]
        };

        let tsc_output = if cfg!(target_os = "windows") {
            Command::new("tsc")
                .args(&tsc_args)
                .current_dir(&temp_dir)
                .output()
        } else {
            Command::new("tsc")
                .args(&tsc_args)
                .current_dir(&temp_dir)
                .output()
        };

        let result = match tsc_output {
            Ok(output) => {
                if !output.status.success() {
                    let error_output = String::from_utf8_lossy(&output.stdout);
                    let stderr_output = String::from_utf8_lossy(&output.stderr);

                    let combined_error = if !error_output.is_empty() {
                        error_output.to_string()
                    } else if !stderr_output.is_empty() {
                        stderr_output.to_string()
                    } else {
                        "TypeScript compilation failed".to_string()
                    };

                    Err(parse_typescript_error(&combined_error, Some(ns_path.to_path_buf())))
                } else {
                    // Move the generated JS file to the target location
                    match fs::metadata(&temp_js_path).await {
                        Ok(_) => {
                            // Create output directory if needed
                            if let Some(parent) = js_path.parent() {
                                fs::create_dir_all(parent).await?;
                            }
                            fs::copy(&temp_js_path, js_path).await?;
                            Ok(())
                        }
                        Err(_) => {
                            Err(NullScriptError::Transpile(NullScriptTranspileError::with_location(
                                "JavaScript file was not generated by TypeScript compiler".to_string(),
                                Some(ns_path.to_path_buf()),
                                None,
                                None,
                            )))
                        }
                    }
                }
            }
            Err(e) => Err(NullScriptError::Io(e)),
        };

        // Clean up temporary directory
        let _ = fs::remove_dir_all(&temp_dir).await;

        result
    }

    /// Build an entire directory of .ns files
    pub async fn build_directory(
        &self,
        input_dir: &Path,
        output_dir: &Path,
        options: &TranspileOptions,
    ) -> Result<Vec<PathBuf>, NullScriptError> {
        let mut outputs = Vec::new();

        for entry in WalkDir::new(input_dir)
            .into_iter()
            .filter_map(|e| e.ok())
            .filter(|e| e.path().extension().map_or(false, |ext| ext == "ns"))
        {
            let ns_file = entry.path();
            let relative_path = ns_file.strip_prefix(input_dir)
                .map_err(|e| NullScriptError::Io(std::io::Error::new(std::io::ErrorKind::InvalidInput, e)))?;

            let output_ext = match options.output_format {
                OutputFormat::JavaScript => "js",
                OutputFormat::TypeScript => "ts",
            };

            let output_path = output_dir.join(relative_path.with_extension(output_ext));

            match options.output_format {
                OutputFormat::JavaScript => {
                    self.transpile_to_js(ns_file, &output_path, options).await?;
                }
                OutputFormat::TypeScript => {
                    self.transpile_file(ns_file, &output_path, options).await?;
                }
            }

            outputs.push(output_path);
        }

        Ok(outputs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[tokio::test]
    async fn test_basic_transpilation() {
        let transpiler = NullScriptTranspiler::new();
        let source = r#"
definitely message = "Hello, World!";
maybe count = 0;
checkthis (count is 0) {
    console.log(message);
}
"#;

        let result = transpiler.transpile(source, &TranspileOptions::default()).unwrap();

        assert!(result.contains("const message"));
        assert!(result.contains("let count"));
        assert!(result.contains("if (count === 0)"));
    }

    #[tokio::test]
    async fn test_function_transpilation() {
        let transpiler = NullScriptTranspiler::new();
        let source = r#"
feels greet(name: string): string {
    pls `Hello, ${name}!`;
}
"#;

        let result = transpiler.transpile(source, &TranspileOptions::default()).unwrap();

        assert!(result.contains("function greet"));
        assert!(result.contains("return `Hello"));
    }

    #[tokio::test]
    async fn test_syntax_validation() {
        let transpiler = NullScriptTranspiler::new();
        let source = "const invalid = 'should fail';";

        let result = transpiler.validate_syntax(source, None);
        assert!(result.is_err());
    }
}
